Replace JSHint with ESLint and the new JavaScript styleguide

bjacobel opened the pull request:
[TNL-5072](https://openedx.atlassian.net/browse/TNL-5072)

### What I did:
- Install ESLint and `eslint-config-edx`, our new Javascript style guide
- Modify existing paver quality tasks to run ESLint ~~instead of~~ as well as JSHint
- Modify existing CI scripts that call paver tasks to reference ESLint ~~instead of~~ as well as JSHint
- Determine the full list of Javascript files that we do not want to lint (because they are vendor files, autogenerated, pre-minified, already gitignored, etc.)
- Upgrade version of `diff-cover` to pick up the fixes I made to the eslint integration there

### What I did not do:
- Run `eslint --fix` on the codebase to autofix simple varieties of lint failures (preliminary investigation shows this could eliminate 40-50% of our linter violations, but I wanted to keep all code changes out of this PR to increase the chances it will land)
- Remove `// jshint ignore:line` directives or similar in the code base (there are 200-ish of these, mostly they can be changed directly to ESLint line ignores and I have another branch off this one where I've started that work)
- Fix any lint violations, anywhere
- Touch any code that runs anywhere besides Jenkins. There are fairly massive code changes coming, but I think it would be best if this was landed and stable before that happens.

### Unresolved questions:
- ~~We now pipe output to `eslint.log` instead of `jshint.log` in CI. (see changes in `generic_ci_tests.sh`.) Will our CI environment need additional changes to pick this change up?~~ **Update**: Yes, it will, I've filed TE tickets for the work that will need to be done there.
- ~~When will [my PR to `diff-cover`](https://github.com/Bachmann1234/diff-cover/pull/46) land? Is it a blocker for this (i.e., is installing from a git URI pointing to my fork not okay to merge to master? Probably.) (**Update** : my PR merged and the maintainer will be cutting a new release tonight!)~~ **Update**: That PR landed in the `diff-cover` 0.9.8 release, which this upgrades to.

### Sandbox
- sandbox doesn't seem neccessary for this as it's devtools-only

### Testing
- ~~i18n~~
- ~~RTL~~
- ~~safecommit violation code review process~~
- [x] Unit, integration, acceptance tests as appropriate
- ~~Analytics~~
- ~~Performance~~
- ~~Database migrations are backwards-compatible~~

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [x] @andy-armstrong 
- [ ] @benpatterson 
- [x] @cpennington 

There are probably other people on Tools that I should tag, can one of Ben/Cale help me figure out who those people are?

### Post-review
- [ ] Squash commits

bjacobel commented:
jenkins run a11y

bjacobel commented:
Ready for review 🎉

bjacobel commented:
Not sure whether a tools or devops person is more suited to answer my questions about renaming the Jenkins log files. @jzoldak, any ideas on this?

bjacobel commented on scripts/all-tests.sh:
```diff
@@ -12,7 +12,7 @@ set -e
 # Violations thresholds for failing the build
 export PYLINT_THRESHOLD=4175
-export JSHINT_THRESHOLD=7550
+export ESLINT_THRESHOLD=64345
```
Not a typo; we are allowing ~10x more violations initially. This is due to the fact that we're being much, much stricter about our JavaScript style with the adoption of the new styleguide.

Forthcoming PRs will reduce this by about half (autofixing) and possibly more (by adding repo-specific rule exceptions based on our analysis of the most commonly violated rules in existing code).

cpennington replied:
Why add exceptions? Having a high number is ok, if it's an accurate reflection of what we want to fix eventually.

bjacobel replied:
I think of line, file or directory-specific rule exceptions as a way to get us to "zero violations" while acknowledging that the rewards of doing certain maintenance tasks (ie - converting all variables to use camelCase) are less than the costs of refactoring that code (ie - breaking all code that calls your refactored file because the API has changed).

But that's based on the premise of getting to zero reported violations at some point, which maybe we're just not going to attempt.

(there's also several _kinds_ of JS in the repo -- some is requireJS/AMD, some is commonJS, some is just magic global variable house of cards stuff -- and not all the rules can apply logically to all the code. Example: we currently have some rules turned off in the Gulp directory of the toolkit, because it can use `'use strict';` differently than our browser code because it runs in Node).

jzoldak commented on scripts/generic-ci-tests.sh:
```diff
@@ -80,8 +80,8 @@ case "$TEST_SUITE" in
-        echo "Finding jshint violations and storing report..."
-        paver run_jshint -l $JSHINT_THRESHOLD > jshint.log || { cat jshint.log; EXIT=1; }
+        echo "Finding ESLint violations and storing report..."
+        paver run_eslint -l $ESLINT_THRESHOLD > eslint.log || { cat eslint.log; EXIT=1; }
```
This will require a change to the jenkins job configuration for the edx-platform-quality jobs for master and PRs in order for the Quality Report (HTML report linked on the left sidebar) to replace the jshint log file with the eslint log file. Can you please put in a TE ticket? We will need to co-ordinate that change whenever this is merged in, keeping in mind how we will handle old branches and new.
It will also require a change to the code that runs the [edX Eng Dashboard](https://edx-eng-dashboard.herokuapp.com/edx). Please put in a TE ticket for that too.

bjacobel replied:
Created [TE-1549](https://openedx.atlassian.net/browse/TE-1549) and [TE-1550](https://openedx.atlassian.net/browse/TE-1550), thanks for the heads up.

cpennington replied:
Can we just make the jobs report both, and for now, the eshint one won't exist, and that will be ok? Then we won't break everyone w/ an outstanding PR when this lands.

bjacobel replied:
Yeah, that might be a good way to approach a switchover. What about `paver run_quality` - should it run diff-quality on both eslint and jshint?

bjacobel replied:
Just pushed a reup of this PR that leaves the JSHint reporting intact for now, with comments to remove it when those two stories land.

jzoldak commented on requirements/edx/base.txt:
```diff
@@ -146,7 +146,7 @@ bok-choy==0.5.3
 coverage==4.0.2
 ddt==0.8.0
-diff-cover==0.9.0
+git+git://github.com/bjacobel/diff-cover.git@bjacobel/eslint#egg=diff-cover
```
Do we have a plan for getting this merged upstream? Who will own the followup?

This should be moved from base.txt to github.txt and follow the pattern there. 
That is, git+https://github.com/... 
Also I think there was some trick to adding a version number so that environments that go back and forth between branches (e.g. devstack) keep proper track of which version of a package is installed in the venv. But I don't myself know what that trick is. I'm pretty sure that @nedbat does.

bjacobel replied:
It was merged a few hours ago, and the maintainer told me he'd be releasing a new version to pypi tonight.

bjacobel replied:
Updated to the newly released diff-cover 0.9.8.

cpennington commented on pavelib/assets.py:
```diff
@@ -384,14 +384,24 @@ def coffeescript_files():
+        coffee_files = ["`{}`".format(coffeescript_files())]
+        js_files = r"`{} | sed 's/\.coffee/\.js/g'`".format(coffeescript_files())
+    else:
+        js_files = r"`{} | sed 's/\.coffee/\.js/g'`".format(*coffee_files)
```
Why are we shelling out for this, rather than just converting `.coffee` to `.js` in python?

    if not coffee_files:
        coffee_files = coffeescript_files()
    
    js_files = [re.sub('coffee$', 'js', file) for file in coffee_files]

bjacobel replied:
We need a list of every coffeescript file (to convert them to javascript) and then we need a list of all the javascript files that were converted in order to add this header to them. Can't get that without shelling out, unfortunately, because you have to actually run the `coffeescript_files` command to find the CS files before changing their extensions to JS - just running that same search for JS files will turn up lots of things that weren't converted from coffeescript.

cpennington replied:
Right, we need to find all of the coffeescript files, and shelling out to `find` is a convenient way to do that. We don't need to replace the file suffixes using `sed`, though, we can do that more simply just in python. And we don't need to call the `coffeescipt_files` function more than once.

bjacobel replied:
Yes,,, sort of - `coffee_files` can be either a list of files, or a inlined bash command to find a list of files. You have to run it through bash first to normalize it to something you could run `re.sub` on first. So to rewrite this using `re.sub` instead of `sed` I'd need to make three separate calls of `sh(cmd())` - one to compile the coffeescript, one to turn `coffee_files` into something I can call re.sub on, and one to add the header to the JS files.

(You're definitely right about calling `coffeescript_files` more than once, though.)

cpennington replied:
Ah, I didn't realized that we were passing in a bash command to get the coffeescript files. That explains some of the way the code is written now. (Although, we should probably use the `$(...)` syntax rather than the `` syntax, since it nests properly. Also, I'm a bit skeptical of taking an arbitrary bash command in the first place, tbh.

bjacobel replied:
Pushing up a cleaned-up version of this now, let me know what you think.

bjacobel replied:
Update: I'm abandoning this approach of adding a header to every coffeescript file in favor of just maintaining a complete list of the `.js` versions of them in the `.eslintignore`. It turns out to not be that many folders/files.

cpennington commented on pavelib/quality.py:
```diff
@@ -485,7 +487,10 @@ def _get_report_contents(filename, last_line_only=False):
+                i = 1
+                while lines[len(lines) - i] == '\n':
+                    i += 1
+                return lines[len(lines) - i]
```
for line in reversed(lines):
        if line != '\n':
            return line

cpennington commented on pavelib/assets.py:
```diff
@@ -384,14 +384,25 @@ def coffeescript_files():
 @task
 @no_help
-def compile_coffeescript(*files):
+def compile_coffeescript(*coffee_files):
```
Can you add documentation to this? Is `coffee_files` a list of files, or a bash command to get a list of files? In the former case, `js_find_cmd` is broken, because the format command it uses only has space for a single value.

cpennington commented on pavelib/assets.py:
```diff
@@ -384,14 +384,36 @@ def coffeescript_files():
-    if not files:
-        files = ["`{}`".format(coffeescript_files())]
+
+    coffee_find_cmd = coffeescript_files()
```
No need to call this if you aren't going to use the results, since it's actually calling `cmd`, which will execute the command.

cpennington replied:
Oh, sorry, I misread. I thought `cmd` executed the command, but I see that it doesn't (you have to pass it to `sh` to execute it).

cpennington commented on pavelib/assets.py:
```diff
@@ -384,14 +384,36 @@ def coffeescript_files():
+        coffee_files = ["$({})".format(coffee_find_cmd)]
+        js_files = js_find_cmd.format(coffee_find_cmd)
+    else:
+        js_files = js_find_cmd.format(*coffee_files)
```
This won't work if there's more than one file in `coffee_files`. The `js_find_command` only has a single slot to format into.

cpennington replied:
The same is true with `coffee_files` in the previous section of the block.

bjacobel commented:
jenkins run all

bjacobel commented:
This is ready for another round of review.

bjacobel commented on .eslintignore:
```diff
@@ -0,0 +1,57 @@
+common/lib/xmodule/xmodule/js/src/sequence/display.js
+common/lib/xmodule/xmodule/js/src/sequence/edit.js
+common/lib/xmodule/xmodule/js/src/tabs/tabs-aggregator.js
+common/lib/xmodule/xmodule/js/src/vertical/edit.js
```
FYI, I've verified that I have all Coffeescript files ignored from eslint by doing a `git clean -fdX`, running ESLint, running `paver compile_coffeescript`, running ESLint, and checking that the violation numbers from the two runs match.

andy-armstrong commented on .eslintrc.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "extends": "eslint-config-edx"
+}
```
I know we'd discussed that there were going to be some overrides but I don't remember what they were now. It makes sense to check this in as you have it, and then we can add overrides if or when we need them in follow-on PRs.

bjacobel replied:
Yeah, I figure this PR will establish the high-water mark for violations and then we can draw it down as we come across things we think should be exceptions in this repo. Cale and I talked about this a bit [here](https://github.com/edx/edx-platform/pull/13081#discussion_r72491949)

andy-armstrong commented on package.json:
```diff
@@ -18,8 +18,8 @@
   },
   "devDependencies": {
     "edx-custom-a11y-rules": "0.1.2",
-    "pa11y": "3.6.0",
```
Thank you for alphabetizing the dependencies!

bjacobel replied:
`npm install --save` does this automatically, I'm more confused how they got *out* of alphabetical order in the first place.

andy-armstrong commented on pavelib/paver_tests/test_eslint.py:
```diff
@@ -8,16 +8,16 @@
 from paver.easy import BuildFailure
 
 
-class TestPaverJsHint(unittest.TestCase):
```
Did you want to keep the JSHint tests since you are now keeping JSHint around? Since we know we're planning on removing it soon then maybe removing it is okay, but it might be better to keep it just in case we don't get around to doing the follow-on work.

andy-armstrong commented on pavelib/paver_tests/test_paver_quality.py:
```diff
@@ -61,7 +61,7 @@ def test_pep8_parser(self):
     """
     For testing utility functions for getting counts from reports for
-    run_jshint, run_complexity, run_safelint, and run_safecommit_report.
+    run_eslint, run_complexity, run_safelint, and run_safecommit_report.
```
As mentioned in the other file, should we keep the JSHint tests until the feature is removed?

andy-armstrong commented:
👍  this ended up being much simpler than I expected. I had a couple of nits on the tests (whether we should keep testing JSHint). I'm happy to merge either way.

I'm in favor of the approach of keeping both linters around to ease the transition of all the quality reports. Is there a PR for removing JSHint once we're ready?

bjacobel commented:
Ok, I've now added back all JSHint tests. There's still a few strings where I just changed eslint to jshint, just because it's already done and it will make the migration process easier when we go back and remove all of jshint. Ready for re-review, @cpennington

cpennington commented:
:+1:

bjacobel commented:
Thanks for the review, @cpennington!

As this affects a lot of people who may not necessarily know about it until their quality builds start failing on Jenkins, I'm going to send an email around to edx-code describing these changes and how to configure your editor to use eslint while you're developing before I merge this.
